### ISSUE TITLE
NoSuchMethodException: java.lang.Class.getDeclaredFields0(boolean)

### ISSUE THREAD

**Issue by jyinpci** [eclipse-openj9/openj9#18788]
Java -version output
--------------------

openjdk version "17.0.9" 2023-10-17
IBM Semeru Runtime Open Edition 17.0.9.0 (build 17.0.9+9)
Eclipse OpenJ9 VM 17.0.9.0 (build openj9-0.41.0, JRE 17 Windows 11 amd64-64-Bit Compressed References 20231017_568 (JIT enabled, AOT enabled)
OpenJ9   - 461bf3c70
OMR      - 5eee6ad9d
JCL      - 3699725139c based on jdk-17.0.9+9)

Summary of problem
------------------

Received the NoSuchMethodException: java.lang.Class.getDeclaredFields0(boolean) when ran the following piece of code in a Springboot application (in Groovy).  This code works fine with Corretto-17 JDK.

            import java.lang.reflect.Method;

            Method getDeclaredFields0 = Class.class.getDeclaredMethod("getDeclaredFields0", boolean.class);
            getDeclaredFields0.setAccessible(true);

Diagnostic files
----------------

Here is the stacktrace:

	java.lang.NoSuchMethodException: java.lang.Class.getDeclaredFields0(boolean)
		at java.base/java.lang.Class.newNoSuchMethodException(Class.java:660)
		at java.base/java.lang.Class.throwExceptionOrReturnNull(Class.java:1458)
		at java.base/java.lang.Class.getMethodHelper(Class.java:1544)
		at java.base/java.lang.Class.getDeclaredMethod(Class.java:1132)
		at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
		at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
		at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
		at java.base/java.lang.reflect.Method.invoke(Method.java:568)
		at org.codehaus.groovy.reflection.CachedMethod.invoke(CachedMethod.java:343)
		at groovy.lang.MetaMethod.doMethodInvoke(MetaMethod.java:328)
		at groovy.lang.MetaClassImpl.doInvokeMethod(MetaClassImpl.java:1348)
		at groovy.lang.MetaClassImpl.invokeMethod(MetaClassImpl.java:1103)
		at groovy.lang.MetaClassImpl.invokeMethod(MetaClassImpl.java:1009)
		at groovy.lang.MetaClassImpl.invokeStaticMethod(MetaClassImpl.java:1590)
		at org.codehaus.groovy.vmplugin.v8.IndyInterface.fromCache(IndyInterface.java:318)
		at com.pci.vz_rest_v4_ir.EbComponentRouteBuilder.configure(RoutesScript.groovy:55)
		at org.apache.camel.builder.RouteBuilder.checkInitialized(RouteBuilder.java:672)
		at org.apache.camel.builder.RouteBuilder.configureRoutes(RouteBuilder.java:618)
		at org.apache.camel.builder.RouteBuilder.addRoutesToCamelContext(RouteBuilder.java:554)
		at org.apache.camel.impl.engine.AbstractCamelContext.addRoutes(AbstractCamelContext.java:1178)
		at org.apache.camel.CamelContext$addRoutes$0.call(Unknown Source)
		at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCall(CallSiteArray.java:45)
		at org.apache.camel.CamelContext$addRoutes$0.call(Unknown Source)
		at com.pci.custebalite.component.EbComponentInitTrait$Trait$Helper.addCustomRoutes(EbComponentInitTrait.groovy:303)
		at org.codehaus.groovy.vmplugin.v8.IndyInterface.fromCache(IndyInterface.java:318)
		at com.pci.vz_rest_v4_ir.InitProc.addCustomRoutes(InitProcScript.groovy)
		at com.pci.custebalite.component.EbComponentInitTrait$addCustomRoutes$3.call(Unknown Source)
		at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCall(CallSiteArray.java:45)
		at com.pci.custebalite.component.EbComponentInitTrait$addCustomRoutes$3.call(Unknown Source)
		at com.pci.custebalite.component.EbComponentInitTrait$Trait$Helper.process(EbComponentInitTrait.groovy:77)
		at org.codehaus.groovy.vmplugin.v8.IndyInterface.fromCache(IndyInterface.java:318)
		at com.pci.vz_rest_v4_ir.InitProc.process(InitProcScript.groovy)
		at org.codehaus.groovy.vmplugin.v8.IndyInterface.fromCache(IndyInterface.java:318)
		at com.pci.vz_rest_v4_ir.InitProcScript.run(InitProcScript.groovy:19)
		at groovy.lang.GroovyShell.evaluate(GroovyShell.java:460)
		at groovy.lang.GroovyShell.evaluate(GroovyShell.java:504)
		at groovy.lang.GroovyShell$evaluate$0.call(Unknown Source)
		at com.pci.custebalite.component.AddComponentsProc.addComponent(AddComponentsProcScript.groovy:131)
		at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
		at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
		at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
		at java.base/java.lang.reflect.Method.invoke(Method.java:568)
		at org.codehaus.groovy.reflection.CachedMethod.invoke(CachedMethod.java:343)
		at groovy.lang.MetaMethod.doMethodInvoke(MetaMethod.java:328)
		at org.codehaus.groovy.runtime.metaclass.ClosureMetaClass.invokeMethod(ClosureMetaClass.java:342)
		at org.codehaus.groovy.runtime.callsite.PogoMetaClassSite.callCurrent(PogoMetaClassSite.java:63)
		at org.codehaus.groovy.runtime.callsite.AbstractCallSite.callCurrent(AbstractCallSite.java:203)
		at com.pci.custebalite.component.AddComponentsProc$_process_closure4.doCall(AddComponentsProcScript.groovy:103)
		at jdk.internal.reflect.GeneratedMethodAccessor24.invoke(Unknown Source)
		at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
		at java.base/java.lang.reflect.Method.invoke(Method.java:568)
		at org.codehaus.groovy.reflection.CachedMethod.invoke(CachedMethod.java:343)
		at groovy.lang.MetaMethod.doMethodInvoke(MetaMethod.java:328)
		at org.codehaus.groovy.runtime.metaclass.ClosureMetaClass.invokeMethod(ClosureMetaClass.java:279)
		at groovy.lang.MetaClassImpl.invokeMethod(MetaClassImpl.java:1009)
		at groovy.lang.Closure.call(Closure.java:418)
		at groovy.lang.Closure.call(Closure.java:434)
		at org.codehaus.groovy.runtime.ResourceGroovyMethods.eachFile(ResourceGroovyMethods.java:1224)
		at org.codehaus.groovy.runtime.ResourceGroovyMethods.eachDir(ResourceGroovyMethods.java:1258)
		at org.codehaus.groovy.runtime.dgm$1057.invoke(Unknown Source)
		at org.codehaus.groovy.runtime.callsite.PojoMetaMethodSite$PojoMetaMethodSiteNoUnwrapNoCoerce.invoke(PojoMetaMethodSite.java:253)
		at org.codehaus.groovy.runtime.callsite.PojoMetaMethodSite.call(PojoMetaMethodSite.java:57)
		at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:139)
		at com.pci.custebalite.component.AddComponentsProc.process(AddComponentsProcScript.groovy:84)
		at org.apache.camel.Processor$process.call(Unknown Source)
		at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCall(CallSiteArray.java:45)
		at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:125)
		at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:139)
		at com.pci.custebalite.StaticRoutes$_configure_closure17.doCall(StaticRoutes.groovy:240)
		at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
		at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
		at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
		at java.base/java.lang.reflect.Method.invoke(Method.java:568)
		at org.codehaus.groovy.reflection.CachedMethod.invoke(CachedMethod.java:343)
		at groovy.lang.MetaMethod.doMethodInvoke(MetaMethod.java:328)
		at org.codehaus.groovy.runtime.metaclass.ClosureMetaClass.invokeMethod(ClosureMetaClass.java:279)
		at groovy.lang.MetaClassImpl.invokeMethod(MetaClassImpl.java:1009)
		at groovy.lang.Closure.call(Closure.java:418)
		at org.codehaus.groovy.runtime.ConvertedClosure.invokeCustom(ConvertedClosure.java:50)
		at org.codehaus.groovy.runtime.ConversionHandler.invoke(ConversionHandler.java:113)
		at jdk.proxy2/jdk.proxy2.$Proxy131.process(Unknown Source)
		at org.apache.camel.support.processor.DelegateSyncProcessor.process(DelegateSyncProcessor.java:65)
		at org.apache.camel.processor.errorhandler.RedeliveryErrorHandler$RedeliveryTask.doRun(RedeliveryErrorHandler.java:818)
		at org.apache.camel.processor.errorhandler.RedeliveryErrorHandler$RedeliveryTask.run(RedeliveryErrorHandler.java:726)
		at org.apache.camel.impl.engine.DefaultReactiveExecutor$Worker.schedule(DefaultReactiveExecutor.java:189)
		at org.apache.camel.impl.engine.DefaultReactiveExecutor.scheduleMain(DefaultReactiveExecutor.java:61)
		at org.apache.camel.processor.Pipeline.process(Pipeline.java:182)
		at org.apache.camel.impl.engine.CamelInternalProcessor.process(CamelInternalProcessor.java:399)
		at org.apache.camel.component.timer.TimerConsumer.sendTimerExchange(TimerConsumer.java:210)
		at org.apache.camel.component.timer.TimerConsumer$1.run(TimerConsumer.java:76)
		at java.base/java.util.TimerThread.mainLoop(Timer.java:573)
		at java.base/java.util.TimerThread.run(Timer.java:523)

**Comment by pshipton** [comment]
Class.getDeclaredFields0() is an internal implementation detail of Hotspot builds and doesn't exist in OpenJ9. It's not part of the Java specification.

**Comment by jyinpci** [comment]
> Class.getDeclaredFields0() is an internal implementation detail of Hotspot builds and doesn't exist in OpenJ9. It's not part of the Java specification.

Thank you very much @pshipton for your quick response!  Just curious, does OpenJ9 have something equivalent to Class.getDeclaredFields0(), so I can workaround it?  I need to use it to override some final fields in HttpURLConnection class, using reflection.

**Comment by babsingh** [comment]
> does OpenJ9 have something equivalent to Class.getDeclaredFields0(), so I can workaround it?

1. The closest native equivalent will be `Field[]  java.lang.Class.getDeclaredFieldsImpl()`, but it doesn't toggle between `public only` and `all` fields.
https://github.com/eclipse-openj9/openj9/blob/b669217a6bfffc51e01fd48d364af4bf120e3c21/jcl/src/java.base/share/classes/java/lang/Class.java#L1231-L1241

2. There is also a public Java API, `Field[] java.lang.Class.getDeclaredFields()`: https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/lang/Class.html#getDeclaredFields().

It will be preferred to use the public API. @jyinpci Do the above solutions work for you?

**Comment by jyinpci** [comment]
@babsingh, thanks for your suggestion.

I tried both but could not accomplish what I needed.

What I need to do is to modify a final field of HttpURLConnection class so that PATCH method can be supported, which can be accomplished by the following code in Hotspot:

            Field methodsField = HttpURLConnection.class.getDeclaredField("methods");
            methodsField.setAccessible(true);

            Method getDeclaredFields0 = Class.class.getDeclaredMethod("getDeclaredFields0", boolean.class);
            getDeclaredFields0.setAccessible(true);
            Field[] fields = (Field[]) getDeclaredFields0.invoke(Field.class, false);

            Field modifiersField = null;
            for (Field each : fields) {
                if ("modifiers".equals(each.getName())) {
                    modifiersField = each;
                    break;
                }
            }
            modifiersField.setAccessible(true);

            // remove the "final" modifier
            modifiersField.setInt(methodsField, methodsField.getModifiers() & ~Modifier.FINAL);

            /* use the following as valid HTTP methods */
            String[] methods = [
                    "GET", "POST", "HEAD", "OPTIONS", "PUT", "DELETE", "TRACE", "PATCH"
            ];
            // set the new methods - including patch
            methodsField.set(null, methods);

When I use getDeclaredFieldsImpl, I received a NoSuchMethodException.  When I use the getDeclaredFields() against the Field class, either explicitly or via reflection, it returns an empty list of Field.

**Comment by babsingh** [comment]
> When I use getDeclaredFieldsImpl, I received a NoSuchMethodException. When I use the getDeclaredFields() against the Field class, either explicitly or via reflection, it returns an empty list of Field.

@jyinpci
- Did you account for the difference in the method signature between `getDeclaredFieldsImpl` and `getDeclaredFields0`?
- Also, you can use `Field.class.getDeclaredField("modifiers")`, which will replace the `for` loop and other related code.

I tried the below and it works:
```
        //Method getDeclaredFields0 = Class.class.getDeclaredMethod("getDeclaredFields0", boolean.class); <-- OLD
        Method getDeclaredFields0 = Class.class.getDeclaredMethod("getDeclaredFieldsImpl"); // <-- NEW
        getDeclaredFields0.setAccessible(true);
        //Field[] fields = (Field[]) getDeclaredFields0.invoke(Field.class, false); <-- OLD
        Field[] fields = (Field[]) getDeclaredFields0.invoke(Field.class); // <-- NEW

        Field modifiersField = null;
        for (Field each : fields) {
            System.out.println("field: " + each.getName()); // <-- Debug
            if ("modifiers".equals(each.getName())) {
                modifiersField = each;
                break;
            }
        }
        System.out.println("modifier field v1: " + modifiersField.getName()); // shows that the field is correctly found
        modifiersField.setAccessible(true);

        System.out.println("modifier field v2: " + Field.class.getDeclaredField("modifiers").getName()); // shows that the field is correctly found
```

**Comment by jyinpci** [comment]
@babsingh, thank you so much for taking the time to answer this!

Yes, getDeclaredFieldsImpl worked - I did not fully switch the JDK to OpenJ9 when I tested it last time.  My bad.

java.lang.reflect.Field.class.getDeclaredField("modifiers") still does not work for me.  I received a java.lang.NoSuchFieldException: modifiers.

But the getDeclareFieldsImpl should be good enough for me now.  Thanks again!
